src/rendering/batcher/shared/Batcher.ts

```diff
@@ -142,27 +169,72 @@ export class Batcher
     public batchIndex = 0;
     public batches: Batch[] = [];
 
-    // specifics.
-    private readonly _vertexSize: number = 6;
-
-    private _elements: BatchableObject[] = [];
+    private _elements: BatchableElement[] = [];
 
     private _batchIndexStart: number;
     private _batchIndexSize: number;
 
     /** The maximum number of textures per batch. */
     public readonly maxTextures: number;
 
+    /** The name of the batcher. Must be implemented by subclasses. */
+    public abstract name: string;
+    /** The vertex size of the batcher. Must be implemented by subclasses. */
+    public abstract vertexSize: number;
+
+    /** The geometry used by this batcher. Must be implemented by subclasses. */
+    public abstract geometry: Geometry;
+
+    /**
+     * The shader used by this batcher. Must be implemented by subclasses.
+     * this can be shared by multiple batchers of the same type.
+     */
+    public abstract shader: Shader;
+
+    /**
+     * Packs the attributes of a BatchableMeshElement into the provided views.
+     * Must be implemented by subclasses.
+     * @param element - The BatchableMeshElement to pack.
+     * @param float32View - The Float32Array view to pack into.
+     * @param uint32View - The Uint32Array view to pack into.
+     * @param index - The starting index in the views.
+     * @param textureId - The texture ID to use.
+     */
+    public abstract packAttributes(
+        element: BatchableMeshElement,
+        float32View: Float32Array,
+        uint32View: Uint32Array,
+        index: number,
+        textureId: number
+    ): void;
+
+    /**
+     * Packs the attributes of a BatchableQuadElement into the provided views.
+     * Must be implemented by subclasses.
+     * @param element - The BatchableQuadElement to pack.
+     * @param float32View - The Float32Array view to pack into.
+     * @param uint32View - The Uint32Array view to pack into.
+     * @param index - The starting index in the views.
+     * @param textureId - The texture ID to use.
+     */
+    public abstract packQuadAttributes(
+        element: BatchableQuadElement,
+        float32View: Float32Array,
+        uint32View: Uint32Array,
+        index: number,
+        textureId: number
+    ): void;
+
     constructor(options: BatcherOptions = {})
     {
         Batcher.defaultOptions.maxTextures = Batcher.defaultOptions.maxTextures ?? getMaxTexturesPerBatch();
         options = { ...Batcher.defaultOptions, ...options };
 
-        const { vertexSize, indexSize, maxTextures } = options;
+        const { maxTextures, vertexInitialSize, indexInitialSize } = options;
 
-        this.attributeBuffer = new ViewableBuffer(vertexSize * this._vertexSize * 4);
+        this.attributeBuffer = new ViewableBuffer(vertexInitialSize * 6 * 4);
```

Comment:
why is this no longer using `vertexSize`?
e.g.
`this.attributeBuffer = new ViewableBuffer(vertexInitialSize * this.vertexSize * 4);`

Reply:
removed this.vertexSize from the multiplicaiton as this is abstract and causing errors. Will leave it for the user to factor in when parsing in `vertexInitialSize`